pkg/certs/util.go

```diff
@@ -0,0 +1,209 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// CASecret defines plugged-in secret name
+	PlugInSecret = "cacerts"
+	// RootCertName defines plugged-in root cert file name
+	RootCertName = "root-cert.pem"
+	// CAKeyName defines plugged-in ca private key file name
+	CAKeyName = "ca-key.pem"
+	// CACertName  defines plugged-in ca cert file name
+	CACertName = "ca-cert.pem"
+	//CertChainName  defines plugged-in cert-chain file name
+	CertChainName = "cert-chain.pem"
+
+	caKeySize = 2048
+)
+
+// PlugInProvisionedCert plugin external CAs into Istio
+
+func PlugInProvisionedCert(client kubernetes.Interface, certDir, ns string, dryRun bool) error {
+	stat, err := os.Stat(certDir)
+	if err != nil {
+		return err
+	}
+	if stat.Mode().Perm() != 0777 {
+		return fmt.Errorf("specified certDir %q should have mode %#o, has %#o", certDir, os.ModePerm, stat.Mode())
+	}
+	rootCert, err := ioutil.ReadFile(filepath.Join(certDir, RootCertName))
+	if err != nil {
+		log.Warnf("can not read %q from %s", RootCertName, certDir)
+		return err
+	}
+	cakey, err := ioutil.ReadFile(filepath.Join(certDir, CAKeyName))
+	if err != nil {
+		log.Warnf("can not read %q from %s", CAKeyName, certDir)
+		return err
+	}
+	caCert, err := ioutil.ReadFile(filepath.Join(certDir, CACertName))
+	if err != nil {
+		log.Warnf("can not read %q from %s", CACertName, certDir)
+		return err
+	}
+	certChain, err := ioutil.ReadFile(filepath.Join(certDir, CertChainName))
+	if err != nil {
+		log.Warnf("can not read %q from %s", CertChainName, certDir)
+		return err
+	}
+
+	err = createSecret(client, ns, rootCert, cakey, caCert, certChain, dryRun)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+//most parts are coming from security/tools/generate_cert/main.go
+// GenerateCertsAndKey is used to generate key and certs
+func GenerateCertsAndKey(client kubernetes.Interface, ns string, dryRun bool, outputDir, org string) error {
+	_, scrtErr := client.CoreV1().Secrets(ns).Get(context.TODO(), PlugInSecret, metav1.GetOptions{})
+	if scrtErr == nil {
+		log.Infof(PlugInSecret+"already exists in %v, skip generating self-signed Certs and Keys", ns)
+		return nil
+	}
+	if dryRun {
+		log.Infof("dry run mode: skip generating self-signed Certs and Keys ")
+		return nil
+	}
+	if org == "" {
+		org = "istio.io"
+	}
+	options := util.CertOptions{
+		TTL:          cmd.DefaultSelfSignedCACertTTL,
+		Org:          org,
+		IsCA:         true,
+		IsSelfSigned: true,
+		RSAKeySize:   caKeySize,
+		IsDualUse:    true,
+	}
+	rootCertBytes, rootKeyBytes, ckErr := util.GenCertKeyFromOptions(options)
+	if ckErr != nil {
+		return ckErr
+	}
+
+	rootCert, err := util.ParsePemEncodedCertificate(rootCertBytes)
+	if err != nil {
+		return err
+	}
+
+	rootKey, err := util.ParsePemEncodedKey(rootKeyBytes)
+	if err != nil {
+		return err
+	}
+	err = saveRootCert(outputDir, rootCertBytes)
+	if err != nil {
+		return err
+	}
+	opts := util.CertOptions{
+		Org:          org + "intermediate",
```

Comment:
What is this string "intermediate" ?

Reply:
Besides generate root-cert, and one intermediate CA cert for use of Istio. We will also generate two intermediate CA for future use of mutli-cluster or VMs. So I added `intermediate` here